Lib/ctypes/_util.py

```diff
@@ -0,0 +1,15 @@
+from sys import maxsize
+
+def _last_version(libnames, sep):
+    def _num_version(libname):
+        # "libxyz.so.MAJOR.MINOR" => [MAJOR, MINOR]
+        parts = libname.split(sep)
+        nums = []
+        try:
+            while parts:
+                nums.insert(0, int(parts.pop()))
+        except ValueError:
+            pass
+        return nums or [maxsize]
```

Comment:
I would prefer to return None if there no version, but replace None with something else *in the caller*.

Reply:
The code for _last_version comes from util.py:
@@ -170,17 +180,6 @@ elif os.name == "posix":

     if sys.platform.startswith(("freebsd", "openbsd", "dragonfly")):

-        def _num_version(libname):
-            # "libxyz.so.MAJOR.MINOR" => [ MAJOR, MINOR ]
-            parts = libname.split(b".")
-            nums = []
-            try:
-                while parts:
-                    nums.insert(0, int(parts.pop()))
-            except ValueError:
-                pass
-            return nums or [sys.maxsize]
-
         def find_library(name):
             ename = re.escape(name)
             expr = r':-l%s\.\S+ => \S*/(lib%s\.\S+)' % (ename, ename)
@@ -199,7 +198,7 @@ elif os.name == "posix":
             res = re.findall(expr, data)
             if not res:
                 return _get_soname(_findLib_gcc(name))
-            res.sort(key=_num_version)
+            res.sort(key=_last_version)
             return os.fsdecode(res[-1])

I am reluctant to change it (already had one reviewer ask why I had even made _util.py).

And, as our way of finding, better, creating what we are looking through is different - I am more reluctant.

    exprs = [r'lib%s\.so\.[0-9]+[0-9.]*' % name,
        r'lib%s_?64\.so\.[0-9]+[0-9.]*' % name]
    for expr in exprs:
        versions = []
        for line in members:
            m = re.search(expr, line)
            if m:
                versions.append(m.group(0))
        if versions:
            return _last_version(versions, '.')
    return None

* The goal was to have a single _last_version() definition for both uses.
But it is is confusing, maybe there should be two versions (remove _util.py, restore num_version() and move (and adjust _last_version() within _aix.py